components/tidb_query/src/rpn_expr/impl_compare.rs

```diff
@@ -207,9 +213,198 @@ pub fn coalesce<T: Evaluable>(args: &[&Option<T>]) -> Result<Option<T>> {
+        if expr_tp != ExprType::Bytes {
+            return Err(type_error(Bytes::EVAL_TYPE, expr_tp));
+        }
+        Ok(Some(val))
+    }
+}
+
+impl Extract for Decimal {
+    #[inline]
+    fn extract(expr_tp: ExprType, val: Vec<u8>) -> Result<Option<Self>> {
+        if expr_tp != ExprType::Float32 || expr_tp != ExprType::Float64 {
+            return Err(type_error(Decimal::EVAL_TYPE, expr_tp));
+        }
+        use crate::codec::mysql::DecimalDecoder;
+        let value = val
+            .as_slice()
+            .read_decimal()
+            .map_err(|_| other_err!("Unable to decode decimal from the request"))?;
+        Ok(Some(value))
+    }
+}
+
+impl Extract for Duration {
+    #[inline]
+    fn extract(expr_tp: ExprType, val: Vec<u8>) -> Result<Option<Self>> {
+        if expr_tp != ExprType::MysqlDuration {
+            return Err(type_error(Duration::EVAL_TYPE, expr_tp));
+        }
+        let n = val
+            .as_slice()
+            .read_i64()
+            .map_err(|_| other_err!("Unable to decode duration from the request"))?;
+        let value = Duration::from_nanos(n, MAX_FSP)
+            .map_err(|_| other_err!("Unable to decode duration from the request"))?;
+        Ok(Some(value))
+    }
+}
+
+pub trait InByHash: Evaluable + Hash + Eq {}
+pub trait InByCompare: Evaluable + Eq {}
+
+impl InByHash for Int {}
+impl InByHash for Real {}
+impl InByHash for Bytes {}
+impl InByHash for Decimal {}
+impl InByHash for Duration {}
+// DateTime requires TZInfo in context, and we cannot acquire it during metadata_ctor.
+// TODO: implemnet InByHash for DateTime.
+impl InByCompare for DateTime {}
+// Implement Hash for Json is impossible, due to equality of Json depends on an epsilon.
+impl InByCompare for Json {}
+
+#[derive(Debug)]
+pub struct CompareInMeta<T: Eq + Hash> {
+    lookup_set: HashSet<T>,
+    has_null: bool,
+}
+
+#[rpn_fn(varg, capture = [metadata], min_args = 1, metadata_ctor = init_compare_in_data::<T>)]
+#[inline]
+pub fn compare_in_by_hash<T: InByHash + Extract>(
+    metadata: &CompareInMeta<T>,
+    args: &[&Option<T>],
+) -> Result<Option<Int>> {
+    assert!(!args.is_empty());
+    let base_val = args[0];
+    match base_val {
+        None => Ok(None),
+        Some(base_val) => {
+            if metadata.lookup_set.contains(base_val) {
+                return Ok(Some(1));
+            }
+            let mut default_ret = if metadata.has_null { None } else { Some(0) };
+            for arg in &args[1..] {
+                match arg {
+                    None => {
+                        default_ret = None;
+                    }
+                    Some(v) => {
+                        if v == base_val {
+                            return Ok(Some(1));
+                        }
+                    }
+                }
+            }
+            Ok(default_ret)
+        }
+    }
+}
+
+fn init_compare_in_data<T: InByHash + Extract>(expr: &mut Expr) -> Result<CompareInMeta<T>> {
```

Comment:
These logic seems very complex. How about extract them to a new file?

Reply:
move the whole `Extract` trait to `components/tidb_query/src/rpn_expr/extract.rs`?

Reply:
Maybe you can move it to `impl_compare_in.rs`?